### ISSUE TITLE
Consider unsafely assuming in narrowing that multiple inheritance won't happen

### ISSUE THREAD

**Issue by RandomBrainCode** [astral-sh/ty#1578]
### Summary

Is it possible for Ty to infer the type of the Iterator Variable when Iterator is properly defined?

In the below example running, `area` is properly defined above

```
		if allowed:
			area: list[Area] = await user.get_allowed_areas()
		else:
			area: Area = await user.area_get()
```

but when I run `ty check`, I get the following
```
error[unresolved-attribute]: Object of type `object` has no attribute `get_model`
   --> app/routers/me.py:113:36
    |
112 |     if isinstance(area, list):
113 |         models: list[AreaModel] = [await a.get_model() for a in area]
    |                                          ^^^^^^^^^^^
114 |         return JSONResponse(content=loads(dumps([a.content() for a in models])), status_code=200)
    |
info: rule `unresolved-attribute` is enabled by default
```

I was able to resolve it by adding `a: Area` above line 113, but that causes ruff to throw warning `Local variable 'a' is annotated but never used`

One thing, I am not sure if this could be impacting it, but the variable `area` can actually be of type `Area` or type `list[Area]`, hence the `if isinstance(area, list):` at line 112.

### Version

ty 0.0.1-alpha.26 (b225fd8b4 2025-11-10)

**Comment by RandomFractal** [comment]
**_CORRECTION_**: Annotating `a: Area` first did not resolve the issue:
```
error[unresolved-attribute]: Object of type `object` has no attribute `get_model`
   --> app/routers/me.py:114:36
    |
112 |     if isinstance(area, list):
113 |         a: Area  # noqa: F842
114 |         models: list[AreaModel] = [await a.get_model() for a in area]
    |                                          ^^^^^^^^^^^
115 |         return JSONResponse(content=loads(dumps([m.content() for m in models])), status_code=200)
    |
info: rule `unresolved-attribute` is enabled by default
```

**Comment by carljm** [comment]
Thanks for the report! What ty is (subtly) telling you here is that it is possible for there to exist a subclass of `Area` that is also a subclass of `list`, for example.

```py
class Hypothetical(Area, list[int]): ...
```

So because of that possibility, `isinstance(area, list): ...` does not actually guarantee that we have a `list[Area]`, it could be a list of anything. That's why ty infers `object` when you iterate over `area` after that check.

One alternative here is to use `not isinstance(area, Area)` in place of `isinstance(a, list)`; this [fixes the issue](https://play.ty.dev/748a6d1d-1b56-4e91-a857-0bec2c9bde43). 

Another alternative, if you don't ever subclass `Area`, is to mark it `@typing.final`, which will allow ty to make various more precise inferences based on the knowledge that it won't be subclassed.

This issue is related to #1566 and #1567 in that they are all examples of how ty is being "correct but pedantic" about narrowing, in a way which users may not prefer or be accustomed to. We do value "correct" highly, but I will leave this issue open as well, with `needs-decision` tag, so we can evaluate whether we need to consider doing something different.

**Comment by carljm** [comment]
Another thing we could try to do here (and in #1567 and similar issues) is add better explanatory diagnostics. This is tricky since these issues tend to manifest as just "unexpected `object` type inferred", but it might be possible by using a special variant of `object` in certain places (e.g. top materialization method return types) and then special-casing diagnostics involving that special `object` type.

**Comment by RandomFractal** [comment]
@carljm Thanks for the reply!

I will give @typing.final a shot in the next couple of days and let you know if that resolves this scenario. The `Area` object will never be subclassed, so doing this should not cause any issues in my particular scenario.

**Comment by carljm** [comment]
> Another thing we could try to do here (and in [#1567](https://github.com/astral-sh/ty/issues/1567) and similar issues) is add better explanatory diagnostics. This is tricky since these issues tend to manifest as just "unexpected `object` type inferred", but it might be possible by using a special variant of `object` in certain places (e.g. top materialization method return types) and then special-casing diagnostics involving that special `object` type.

I opened #2215 to track this idea.

**Comment by cscutcher** [comment]
@carljm thanks for clarifying and for the hard work on `ty`!

I'm a bit worried that as things stand, and without relying on `ty_extensions`, that this behaviour makes `ty` difficult to use in a lot of common situations. Consider;

```python
Validator = Callable[...]
Validators = Collection[Validator]
MaybeValidators = Validator | Validators
```

If I want to handle a variable of type `MaybeValidators` I think, based on this and other similar issues (apologies if I'm wrong!), I'd need to do.

```python
var: MaybeValidators
if not isinstance(var, Collection):
    # I'm a validator
elif not callable(var):
    # I'm a collection of validators.
else: 
    # I could still be both! So I could still end up here, and I can't do
    # assert_never(var)
	raise ...    
```

I think the possibility that `MaybeValidators` could be a callable that's also a collection is definitely interesting and important to note, but if it requires writing the code like this if-statement to be accurate and valid, I think that might hold `ty` back in terms of adoption.

This is especially true in existing codebases where designing from scratch to avoid situations like this is less plausible. I can imagine submitting a PR where not everyone on the team is using `ty`, and code like this would be rejected for being unnecessarily convoluted.

I'd love to see some of the enhancements that are in `ty_extensions` make it into base python so that more nuanced declaration of types was possible, but I'm sure that's not easy. Again, it's great `ty_extensions` is an option, but a tight coupling to a specific type checker for a project is a tough sell.

On the other hand, it is nice that being this strict is an option, but I think it really does need to be an option that can be turned down or off without neutering other checks.

**Comment by carljm** [comment]
Hi @cscutcher -- thanks for the thoughtful feedback!

It is true that ty asks you to decide what you want to do in the case where it's both, or else explicitly eliminate the possibility by using `typing.final` on one of the types involved. In most cases, where one of the two types involved is a concrete type that can be fully confirmed by a runtime check, you can handle this union naturally with just an if/else, by checking for the concrete type. (That is, if in your example `Validator` were instead `class Validator: ...`, all you would need is `if isinstance(var, Validator): ... else: ...` and everything would work as expected.)

The case you present is the most challenging one, because both `Validator` and `Collection[Validator]` are types that can only be excluded by a supported type-narrowing check, not confirmed. (We can confirm something is a `Collection` but not that it's a `Collection[Validator]`; we can confirm something is `Callable` but not its signature.) This is why you have to use the `if not` construction.

I think in this particular case, we already have the option you need to allow you to ignore the problem. If you disable `call-top-callable` diagnostic, ty won't mind if you just check `if callable(var):` and then call `var` with any set of arguments; this is similar to the behavior of other type checkers.

I'm not closed to the idea that there's more we should do here, but it's not clear to me yet exactly what.

**Comment by cscutcher** [comment]
@carljm thanks for your reply!

If I could expand on previous example, imagine I want to do something like this;

```python
type Validator = Callable[[str], None]
type Validators = Collection[Validator]
type MaybeValidators = Validator | Validators


def normalise_validators(inp: MaybeValidators) -> Validators:
    if isinstance(inp, Collection):
        return inp
    if callable(inp):
        return [inp]
    assert_never(inp)
```

as far as I can tell `ty` seems to fail on this when mypy and pyright don't. What's the best ty friendly way to write this, that doesn't obscure the original intent? 

I'd also be interested if there's a way to write this more correctly with the extensions in `ty_extensions`. I just tried but failed miserably!

**Comment by carljm** [comment]
@cscutcher What behavior would you want to see if someone passes you an object that is both callable and a collection?

Here's probably the simplest option:

```py
def normalise_validators(inp: MaybeValidators) -> Validators:
    if isinstance(inp, Collection):
        assert not callable(inp)
        return inp
    else:
        return [inp]
```

Here's a version of the `MaybeValidators` type using intersections to express that we don't want to accept anything that is both callable and a collection, which works with your initial version of `normalise_validators`:

```py
from ty_extensions import Intersection, Not

type Validator = Callable[[str], None]
type Validators = Collection[Validator]
type MaybeValidators = Intersection[Validator, Not[Collection[object]]] | Intersection[Validators, Not[Callable[..., object]]]
```

That would be a bit more readable if we had syntax support: `(Validator & ~Collection[object]) | (Validators & ~Callable[..., object])`.

But the bigger problem with it is that asking ty to prove that arguments cannot be both will make it difficult to call the function; you won't be able to call it with a list of functions, because the `list` type includes arbitrary subclasses of list, which might be callable.

**Comment by carljm** [comment]
To be clear: I do think we need a better pragmatic answer for cases like this than we currently have; I'm just not sure yet what that should be. I don't love the idea of introducing a global "assume everything is disjoint like other type checkers do" flag.

**Comment by carljm** [comment]
Here's a better option, if we are willing to use intersections:

```py
from typing import Callable, Collection

from ty_extensions import Intersection, Not

type Validator = Intersection[Callable[[str], None], Not[Collection[object]]]
type Validators = Collection[Validator]
type MaybeValidators = Validator | Validators


def normalise_validators(inp: MaybeValidators) -> Validators:
    if isinstance(inp, Collection):
        return inp
    else:
        return [inp]

def v1(x: str) -> None: ...
def v2(x: str) -> None: ...

normalise_validators([v1, v2])
normalise_validators(v1)

def _(v3: Validator):
    normalise_validators(v3)
    normalise_validators([v3])
```

This splits the difference, by saying that we want to ensure validators are not also collections, but we don't care if a collection is also callable. That makes the typical use cases (where validators are functions) work fine, because ty is confident that a function cannot also be a collection. It still means that if some validators are other callable objects, we need to make those classes final.

**Comment by cscutcher** [comment]
@carljm 
Thanks for taking the time to work through the example. All your suggestions look quite reasonable! I was trying to work out why I didn't stumble onto the first one. 

I tried it locally and I do still get an error with your suggested solution, which is weird because I don't via the playground. 

```
error[invalid-return-type]: Return type does not match returned value                                                                                                                                                                                                                                                                                                                          
  --> example.py:49:51                                                                                                                                                                                                                                                                                                                                                                         
   |                                                                                                                                                                                                                                                                                                                                                                                           
47 | type MaybeValidators = Validator | Validators                                                                                                                                                                                                                                                                                                                                             
48 |                                                                                                                                                                                                                                                                                                                                                                                           
49 | def normalise_validators(inp: MaybeValidators) -> Validators:                                                                                                                                                                                                                                                                                                                             
   |                                                   ---------- Expected `Validators` because of return type                                                                                                                                                                                                                                                                                 
50 |     if isinstance(inp, Collection):                                                                                                                                                                                                                                                                                                                                                       
51 |         assert not callable(inp)                                                                                                                                                                                                                                                                                                                                                          
52 |         return inp                                                                                                                                                                                                                                                                                                                                                                        
   |                ^^^ expected `Validators`, found `(((str, /) -> None) & Collection[object] & ~(() -> object)) | (Collection[Validator] & ~(() -> object))`                                                                                                                                                                                                                                 
53 |     else:                                                                                                                                                                                                                                                                                                                                                                                 
54 |         return [inp]                                                                                                                                                                                                                                                                                                                                                                      
   |                                                                                                                                                                                                                                                                                                                                                                                           
info: rule `invalid-return-type` is enabled by default                                                                                                                                                                                                                                                                                                                                         
```

I'm running `ty 0.0.1-alpha.32` locally, so can only assume there's a difference in version affecting me?


> I don't love the idea of introducing a global "assume everything is disjoint like other type checkers do" flag.

I can definitely see why such a flag is a bad idea, and I like the idea of being more specific in general, since the Python type system can be frustratingly limited in what it's able to express, so it's good to see thought going into making it more accurate and descriptive.

**Comment by carljm** [comment]
> I'm running `ty 0.0.1-alpha.32` locally, so can only assume there's a difference in version affecting me?

Definitely! That version was released Dec 5, which makes it quite old in ty terms, since ty is very actively developed! There have been 18 releases since then; the latest is 0.0.16. And there will probably be a 0.0.17 tomorrow :)

**Comment by carljm** [comment]
Note for future: this is causing us to fail this line in the conformance suite: https://github.com/python/typing/blob/ffd520acd116aacbbba793c8f1f1fd51dfda83ae/conformance/tests/narrowing_typeis.py#L35

**Comment by AlexWaygood** [comment]
> Note for future: this is causing us to fail this line in the conformance suite: https://github.com/python/typing/blob/ffd520acd116aacbbba793c8f1f1fd51dfda83ae/conformance/tests/narrowing_typeis.py#L35

But narrowing and intersections are all unspecified, so I think that assertion may be overfitted to the behaviour of other, more established type checkers. Rather than `x: int = await val`, I think it could probably just as well be `x = await val` and still test what it's meant to be testing

**Comment by m-u-xyz** [comment]
Given that both *mypy* and *pyright* are happy to disregard the "actual objects can be multiply-inherited-both" (or ducktyped-both, or type-regstered-both, or …) problem, maybe a mode switch ("compatible" or "correct") is in order.

Don't get me wrong, I **really** like the fact that ty goes for correctness first here, but when you let it loose on existing code you might want to be somewhat pragmatic.

Idea: maybe we can go for compatible iff ty sees no evidence in the codebase that such an object might actually exist …?

**Comment by carljm** [comment]
> maybe we can go for compatible iff ty sees no evidence in the codebase that such an object might actually exist

One problem here is that making local behavior dependent on a scan of the entire codebase destroys LSP server incrementality. Opening just one file in a large codebase now requires a full codebase scan, and changing any file now potentially changes type checking diagnostics anywhere. ty is carefully designed to maximize incrementality, so this kind of approach is not really feasible for us.

(The other problem is that some codebases are libraries, not applications. For a library it's not a reasonable assumption that calls to a function will exclusively use classes seen in the library.)

But yes, I think some kind of compatibility option might be in order here. It might literally just need to be an "assume unions are disjoint" flag.

**Comment by andythomas** [comment]
I would like to put my 2 cents in, because I encountered the same issues. First, I was confused (and maybe a little annoyed ;) about the info `ty` gave me. But now, I do appreciate the preciseness. Please keep this as an option or even as the default.

1. It served an educational purpose, I used the mouse-over of the LSP function of ty and read about it until I understood it (at least much better than before). 
2. The code I will show later is part of a software library for data acquisition and visualization during the measurements. I can only imagine, what uncommon design patterns some people will use when using our library. Therefore, edge case complaints by the type-checker are much welcomed.

I do not see how to solve this (s. below) without negation/intersection. I was very surprised to learn that Python typing is not like set theory. However, I do not really know other languages and assumed it always is...

I do not know if/how to minimize this while still presenting a somewhat reasonable case. Therefore, if anyone is interested, please find the complete code (40 lines) in the [playground](https://play.ty.dev/2219c2a4-7633-4bc3-af21-d812ecfdc9c5) and after this post. There is the error in line 30, but there is also the debug variable in line 37 that shows me that I covered all cases. 

```python3
import logging

from collections.abc import Callable
from typing import Any

ConfigScheme = tuple[str, tuple, dict[str, Any]]
ConfigValue = str | Callable[[], Any] | ConfigScheme
ConfigParameter = dict[str, ConfigValue]

logger = logging.getLogger(__name__)


def device_query(
    device_handle: VisaDevice | Instrument, config_params: ConfigParameter
) -> dict[str, dict[str, Any]]:
    retquery = {}
    for k, q in config_params.items():
        try:
            if isinstance(q, str):
                try:
                    attr = getattr(device_handle, q)
                except AttributeError:
                    line = str(device_handle.query(q))
                else:
                    if callable(attr):
                        line = attr()
                    else:
                        line = attr
            elif callable(q):
                line = q()
            elif isinstance(q, tuple):
                method = getattr(device_handle, q[0])
                if not callable(method):
                    raise ValueError(f"config_params: method '{q[0]}' is not callable")
                line = str(method(*q[1], **q[2]))
            else:
                q
        except Exception:
            logger.exception("exception during config query.")
            raise
        retquery[k] = line
    return retquery
```

**Comment by carljm** [comment]
@andythomas You did cover all cases, but the issue is the "overlap" -- objects which are both tuples and callables. Right now those objects get handled by whichever of the "callable" or "tuple" branches comes first, and neither branch is prepared to handle them, because a tuple that is also a callable might not be callable with your expected signature, or might not be a tuple with your expected element types.

One fix is to use `elif callable(q) and not isinstance(tuple, q):` and then `elif isinstance(q, tuple) and not callable(q):` -- that excludes the "both" case from either branch. It also means you no longer handle all cases; such objects would then reach the final `else:` branch.

The fundamental question here (which ty can't answer) is how you _want_ to handle such objects.

**Comment by andythomas** [comment]
@carljm Thank you for the help. What I do not understand on a fundamental level is why `elif callable(q) and not isinstance(q, tuple):` followed by an unchanged `elif isinstance(q, tuple):` narrows to `tuple[str, tuple[Unknown, ...], dict[str, Any]] | ((() -> Any) & tuple[object, ...])` and not just `tuple[str, tuple[Unknown, ...], dict[str, Any]]`. This leads to error messages. If I think of set theory, how can the tuple possibly be of another kind?

This would be how I believe handling the objects would be most reasonable.

**Comment by carljm** [comment]
@andythomas The type `((() -> Any) & tuple[object, ...])` represents an object which is both a tuple (of some kind) and a `Callable[[], Any]`. Such an object can exist, because `tuple` can be subclassed, and such a subclass can add a `__call__` method. This object is not necessarily a `ConfigScheme` tuple! Since it is a `Callable[[], Any]`, it can be any kind of tuple at all, and it will still inhabit the `ConfigValue` type, thanks to its callable signature. Just because the `ConfigValue` union includes a tuple type (`ConfigScheme`), that doesn't mean that every tuple which is a `ConfigValue` must be a `ConfigScheme` -- an object which satisfies the `Callable` part of `ConfigValue` could _also_ be a tuple (of any shape at all).

If `q` is such an object, it would not take the `elif callable(q) and not isinstance(q, tuple):` branch, because it would fail the `not isinstance(q, tuple)` check. It would take the subsequent `elif isinstance(q, tuple)` branch. And that's why you get errors -- because we can't actually say with confidence what shape of tuple it is. (Unless you also add the `and not callable(q)` that I suggested, which rules out this "both" object and sends it on the final `else` branch.)

**Comment by sharkdp** [comment]
We recently discussed this internally, and we currently do not plan to change anything fundamental for this very general case of *"I have a union of `A | B`, I want to handle cases `A` and `B` but I don't want to deal with the possibility of `A & B` being non-empty"*.

Note that there are related (more narrowly focused) issues where we *do* plan to make a change, namely to allow unsafe narrowing for equality comparisons with `int`/`str`/`bytes` (see https://github.com/astral-sh/ty/issues/1566).

We also plan to experiment with a better solution for cases involving `isawaitable`/`iscallable` narrowing (see https://github.com/astral-sh/ty/issues/3375).

But for the general case, we ask users to restructure — and thereby fix! — their code, by using one of the discussed techniques:

* **Changing the order of the checks.** Sometimes you can make a precise runtime check on `B`, but not on `A`. In this case, check on `B` first, which leaves you with `A & ~B` for the else case, which you can typically treat like an `A`:

	```py
	class B: ...
	
	# this does not work (and is technically wrong)
	def f(obj: Collection[B] | B) -> Collection[B]:
	    if isinstance(obj, Collection):
	        return obj
	    else:
	        return [obj]
	
	# this is fine
	def f(obj: Collection[B] | B) -> Collection[B]:
	    if isinstance(obj, B):
	        return [obj]
	    else:
	        return obj
	```
	https://play.ty.dev/6279121d-93cc-42a4-8989-6f9551f4fdcb
* **Add additional checks**: if you want to / need to keep the order of original version (`f` above), consider adding additional checks that exclude the other variant:
	```py
	# this is also fine
	def f(obj: Collection[B] | B) -> Collection[B]:
	    if isinstance(obj, Collection) and not isinstance(obj, B):
	        return obj
	    else:
	        return [obj]
	```
    https://play.ty.dev/9b4eb39a-0959-4ab3-a15a-8d85d9dad1a9
* **Forbid subclassing**: By marking one of `A` or `B` as `@final` (assuming one of these types refers to a class, and assuming you have control over that class), you exclude the possibility of common subclasses of `A` or `B`. This makes `A` and `B` disjoint (i.e. the intersection `A & B` empty):

	```py
	@final
	class B: ...
	
	# this is also fine
	def f(obj: Collection[B] | B) -> Collection[B]:
	    if isinstance(obj, Collection):
	        return obj
	    else:
	        return [obj]
	```
    https://play.ty.dev/185d1af0-39cb-4a9c-8b88-36d4bc04dfc6
* **Using intersection types**: (I actually couldn't get this to work. The examples given above do not seem to work anymore?)
* **Use a suppression comment or `cast`**: your code is doing something that is (at least technically) unsafe, so if nothing else helps, consider using a suppression comment or a `cast`:
	```py
	def f(obj: Collection[B] | B) -> Collection[B]:
	    if isinstance(obj, Collection):
	        # common subclasses of `B` and `Collection` will take this
	        # branch, so `obj` is a collection, but not necessarily a
	        # collection of `B`s. We ignore this possibility here.
	        return obj  # ty:ignore[invalid-return-type]
	    else:
	        return [obj]
	```
    https://play.ty.dev/838a0c03-7d1a-4867-bad1-66f0652a8992


Given that there is nothing really actionable here, I'm going to close this ticket. But please feel free to reopen it if you feel strongly that we should reconsider our decision.

**Comment by carljm** [comment]
I think #3375 would also apply to many of the cases discussed here that don't involve `isawaitable` or `iscallable` specifically -- it applies to the resulting types from any `isinstance` check also.

**Comment by carljm** [comment]
https://github.com/astral-sh/ty/issues/3477#issuecomment-4608424288 has an argument that we should reconsider here. But the impact here might look quite different after we stop top materializing by default, so we should revisit then.

**Comment by sharkdp** [comment]
@carljm should this be re-opened, based on your comment in https://github.com/astral-sh/ruff/pull/27308#discussion_r3722333034, and based on some experiments that Charlie was working on?

**Comment by andythomas** [comment]
> ty is being "correct but pedantic" about narrowing, in a way which users may not prefer or be accustomed to. We do value "correct" highly

While this was not obvious to me at first, I think the quote is a very good guideline. It was very educational and, in the end, just adding `@final` solves these issues in 95% of my cases. For the 5%, I actually need `ty` to tell me that I forgot something ;)

**Comment by carljm** [comment]
I keep forgetting that we closed this. I do think (as @andythomas says) that in many (most?) cases, it's not hard to restructure the code to make it satisfy ty and be more correct. What bugs me here is that there clearly are cases that can't be represented in a way that makes ty happy: cases like the one @cscutcher showed above, where both sides of the union are not runtime checkable (e.g. a callable or generic type).

It is also true that those cases are inherently unsound, so if you are aiming for soundness you might want ty to alert you to those cases! And it could be considered a positive that they require a suppression comment, or an alternative workaround like a `TypeGuard`.

It does feel unclear to me what justifies adding the "gradual generic isinstance narrowing mode" we just landed, that wouldn't also justify a "compatility mode" here. So I definitely don't feel fully settled about our decision to close this.
